.github/workflows/ci.yml

```diff
@@ -33,28 +29,47 @@ jobs:
   pre-commit:
     uses: ./.github/workflows/step_pre-commit.yml
 
-  codeql:
-    needs: [pre-commit]
+  static-analysis:
+    needs: [ pre-commit ]
     uses: ./.github/workflows/step_static-analysis.yml
+    permissions:
+      contents: read
+      security-events: write
 
   test-pip:
-    needs: [codeql]
+    needs: [ pre-commit ]
     uses: ./.github/workflows/step_tests-pip.yml
+    secrets:
+      CODECOV_TOKEN: ${{ secrets.CODECOV_TOKEN }}
 
-  test-unit-functional-integration:
-    needs: [codeql]
-    uses: ./.github/workflows/step_test_unit_functional.yml
+  coverage:
+    needs: [ test-pip ]
+    uses: ./.github/workflows/step_coverage.yml
+    secrets:
+      CODECOV_TOKEN: ${{ secrets.CODECOV_TOKEN }}
 
   test-conda:
-    needs: [codeql]
+    needs: [ test-pip ]
     uses: ./.github/workflows/step_tests-conda.yml
+    secrets:
+      CODECOV_TOKEN: ${{ secrets.CODECOV_TOKEN }}
 
   test-ui:
-    needs: [codeql]
+    needs: [ test-pip ]
     uses: ./.github/workflows/step_tests-ui.yml
 
   build:
-    needs: [test-pip, test-conda, test-unit-functional-integration, test-ui]
```

Comment:
What is the rationale behind this dependency structure? This will increase overall time of CI, right?!

Reply:
Depends on how Github starts up the workflows. The rationale is to have in a few stages to catch most likely errors:
- `pre-commit`: style issues and common mistakes
- `test-pip` run the main tests
- `static-analysis`: has overlap with pre-commit so should be run after it. Otherwise can be run anywhere, and since it takes time, makes sense to be in parallel with `test-pip`
- other tests pick up minor issues if any (coverage only splits the tests into flags for codecov to report). They can be run after `test-pip`
- `build`: technically can be anywhere but it looks better at the end because it mirrors the CD workflow (beinng part of the deploy stage)

Reply:
We can run `test-conda` in parallel with `test-pip`, right?! I mean they are the ones that take an eternity to complete. Same goes with `test-ui`, which are relatively fast but can overlap with `test-pip` as well.

Reply:
Indeed, but it is much less likely to catch errors in those two rather than `test-pip`. If we dedicate more to the latter it will more likely:
- pick up the failing job-matrix sooner
- end the whole CI early and send notification

Reply:
Or we can use an action like [this](https://github.com/andymckay/cancel-action) to cancel all other parallel jobs when one fails (which in our case `test-pip`) and we save resources. In a successful workflow case, we save time by running all of them in parallel?

Reply:
Time wise, I don't think it will save much after adding sphinx-builder matrix, and maybe other smaller jobs.

About the action though, the difficulty is striking a balance between `fail-fast` and cancel everything. Sometimes you want to know all failing tests, but some tests you know it will fail the same in the other jobs so you want to end early.

So I've tried to consider the overlap of the test jobs and the number of jobs when distributing the jobs on the workflow.

Reply:
But about the testing time in general, at first glance I see places that can be improved by avoiding writing to files, e.g. via [`pytest-regression`](https://pytest-regressions.readthedocs.io/en/latest/api.html). Also running some profiler on the pytest might be useful for this kind of organization as well.

I'm also considering if the functional+external tests are sufficient for the `test-conda` :thinking:

Reply:
If you could take a look into tests and improve their efficiency, that would be great. Currently they are taking a while to finish. Cheers!!

Reply:
Ok, I've found something obvious, we are not running the tests in parallel :facepalm:. Other than that `external` tests seem to be the most expensive there